Added Support for App & No Authentication

ilamp opened the pull request:
- This adds support for For app & no authentication.
- Added/Updated PHPDocs

ilamp commented:
@freekmurze Will you be able to review this? If you require any changes I can go ahead and make those. Thank you!

freekmurze commented on src/Client.php:
```diff
@@ -16,41 +16,63 @@
-        $this->accessToken = $accessToken;
+        if (is_array($accessTokenOrAppCredentials)) {
+            [$this->appKey, $this->appSecret] = $accessTokenOrAppCredentials;
+        } elseif (is_string($accessTokenOrAppCredentials)) {
```
I really don't like `else` statements. Use an second `if` statement.

freekmurze commented on src/Client.php:
```diff
@@ -661,11 +762,23 @@ public function setAccessToken(string $accessToken): self
+            $auth = [
+                'Authorization' => 'Basic '.base64_encode("{$this->appKey}:{$this->appSecret}"),
+            ];
+        } elseif (isset($this->accessToken)) {
```
`else` is the bane of my existence. Refactor this in dedecated functions `getHeadersForCredentials` and `getHeadersForBearerToken`. Use a ternary for determining which of the functions you need.

ilamp commented:
Thank you @freekmurze. Requested changes are in.

FYI: I'm also considering adding few more exception types. Please let me know if that's okay.

ilamp commented:
@freekmurze Will you be able to merge my pull request sometime soon? Please let me know if I can do anything to help.

freekmurze commented on src/Client.php:
```diff
@@ -16,41 +16,64 @@
 {
-    const THUMBNAIL_FORMAT_JPEG = 'jpeg';
-    const THUMBNAIL_FORMAT_PNG = 'png';
+    public const THUMBNAIL_FORMAT_JPEG = 'jpeg';
```
Please don't add `public` here.

ilamp replied:
Removed the access modifiers. I can open a new pull request for this later

freekmurze commented on src/Client.php:
```diff
@@ -16,41 +16,64 @@
+    public const UPLOAD_SESSION_APPEND = 1;
 
-    /** @var string */
+    /**
```
Leave the formatting as it is.

ilamp replied:
Reverting changes

freekmurze commented on src/Client.php:
```diff
@@ -64,6 +87,11 @@ public function __construct(string $accessToken, GuzzleClient $client = null, in
      *
      * @link https://www.dropbox.com/developers/documentation/http/documentation#files-copy_v2
+     *
+     * @param string $fromPath
```
Do not add docs block where everything is already type hinted.

ilamp replied:
Removed

freekmurze commented on src/Client.php:
```diff
@@ -64,6 +87,11 @@ public function __construct(string $accessToken, GuzzleClient $client = null, in
+     * @param string $fromPath
+     * @param string $toPath
+     * @return array
+     * @throws Exception
```
Do not add exception docblocks

ilamp replied:
Removed

freekmurze commented on src/Client.php:
```diff
@@ -683,11 +785,37 @@ public function setAccessToken(string $accessToken): self
     protected function getHeaders(array $headers = []): array
     {
-        return array_merge([
+        $auth = []; // defaults no-auth, can be used for public endpoints
```
Remove the comment

ilamp replied:
Removed the comment.

ilamp commented:
- Added a couple of quick tests
- Removed the unrelated changes

Thank you for reviewing the pull request!

> Thanks for your work on this.
> 
> This PR contains some unwanted changes. Only PR the bare necessities to make the new functionality work. Could you also add tests?

freekmurze commented:
Thanks!
